### DISCUSSION TITLE
swissTLM3D: Strukturierte Metadaten?

### DISCUSSION THREAD

**Discussion by ratnanil** [GeoUnconference/discussions#29] [Ideen / Idées]
Ich war an der GeoUnconference nicht dabei, ich hoffe ich kann trotzdem ein Thema ansprechen. Zudem weiss ich nicht ob mein Anliegen zu spezifisch ist.. 

Swisstopo Stellt eine ganze Palette an Vektor basierten [Landschaftsmodellen](https://www.swisstopo.admin.ch/de/geodata/landscape.html) zur Verfügung:

- swissTLM3D
- swissTLMRegio
- swissBUILDINGS3D 2.0
- swissBOUNDARIES3D
- swissNAMES3D
 
Viele (alle?) diese Vektordaten haben relativ umfangreiche Attributtabellen, welche die Objekte beschreiben. Beispielsweise beschreibt das Attribut `objektart` im Datensatz `TLM_Strasse` (swissTLM3D) den Strassentyp, welcher als Integer codiert ist. Um heraus zu finden, welcher Code welchen Strassentyp beschreibt muss man den zugehörigen Objektkatalog konsultieren: Eine PDF Datei wo zu jedem Datensatz eine Tabelle enthalten ist (s.u.). Diese Zuordnung muss vom Nutzer manuell erfolgen, diese Tabelle wird nur als PDF angeboten. 

Die Daten liegen der Swisstopo sicher strukturiert vor: **Wie könnte man sich dafür stark machen, dass diese auch entsprechen strukturiert (mindestens als CSV) angeboten werden?**

![DeepinScreenshot_select-area_20210915083112](https://user-images.githubusercontent.com/12532091/133382292-b69393eb-5129-4caa-a809-b47906834aab.png)

PS: Die Tabelle aus dem PDF zu extrahieren stellt sich als eine ziemliche Herausforderung dar, da in einer Tabelle mehrere Attribute beschrieben werden.

**Comment by rastrau** [comment]
Welches Datenformat beziehst Du?

**Reply by ratnanil** [reply]
oh, gibt es da Unterschiede? Leider stehen ja nicht so viele zur Auswahl (gdb, shp, xtf), dabei beziehe ich jeweils .gdb.

**Reply by rastrau** [reply]
Und liest das in R ein oder nutzt es in QGIS?

**Reply by rastrau** [reply]
In Interlis (.xtf) sind die Werte für z.B: `OBJEKTART` explizit enthalten, vgl. Ausschnitt aus dem .ili:

![tlm](https://user-images.githubusercontent.com/4510815/133386930-e74bb60e-4c35-4327-bcf6-985b42321f6a.png)

In der Esri File Geodatabase (.gdb) ist `OBJEKTART` als «coded value domain» modelliert (ArcGIS zeigt direkt die sprechenden Werte an; Du "siehst" in Deinem Workflow wohl nur die Codes links):

![tlm2](https://user-images.githubusercontent.com/4510815/133386931-63c993d8-b33b-416b-a7da-6a31a197da73.png)

Im Shapefile ist `OBJEKTART` direkt mit den "sprechenden" Werten belegt:

![tlm3](https://user-images.githubusercontent.com/4510815/133386932-d30df3c0-a1d5-49d6-b326-068db03e33da.png)

Eventuell kann die von Dir genutzte Software diese Informationen nicht korrekt verarbeiten?

**Reply by ratnanil** [reply]
> Und liest das in R ein oder nutzt es in QGIS?

ich pflege eine lokale Postgres / PostGIS instanz und greife von allem möglichen darauf zu (`R`, `gdal`, QGIS..)

> In Interlis (.xtf) sind die Werte für z.B: OBJEKTART explizit enthalten
> In der Esri File Geodatabase (.gdb) ist OBJEKTART als «coded value domain» modelliert 

"Coded value domains" kann PostGIS nicht interpretieren, aber mit [ili2db](https://github.com/claeis/ili2db) könnte ich allenfalls das Interlis Format verwenden. Werde das auf jeden Fall mal versuchen... vielen Dank!

**Reply by ratnanil** [reply]
I just had to update my TLM Dataset and took the chance to use the interlis data instead of the gdb files. I wrote down the steps necessary in case anyone needs this as well. I'm not sure if my `ili2pg` call is perfect, this is my first time using it. If anyone has feedback on this, feel free!


### Preperation (one time setup)

1. Download [ili2db](https://github.com/claeis/ili2db#download-ili2db) and navigate to the folder
2. Install [the extension](https://github.com/claeis/ili2db/blob/master/docs/ili2db.rst#laufzeitanforderungen) `CREATE EXTENSION "uuid-ossp";`

### Download data and import


1. Download Data from swisstopo (the newest files as interlis)
    - https://www.swisstopo.admin.ch/en/geodata/landscape/tlm3d.html
    - https://www.swisstopo.admin.ch/en/geodata/landscape/boundaries3d.html
2. unzip the files (better to use bash here, since the gui doesn't show the progress)

        unzip swisstlm3d_2021-04_2056_5728.xtf.zip 

3. run the following commands (for each xtf file):

        java -jar ili2pg-4.6.1.jar --import --doSchemaImport --dbhost localhost --dbport 5432 --dbdatabase swisstopo --dbschema swisstlm3d --dbusr postgres --dbpwd HIDDEN --defaultSrsCode 2056  path/to/xtf-file.xtf
  
4. or run the following for loop:

        for file in /path/to/folder/*.xtf; 
        do 
            echo processing $file; 
            java -jar ili2pg-4.6.1.jar --import --doSchemaImport --dbhost localhost --dbport 5432 --dbdatabase swisstopo --dbschema swisstlm3d --dbusr postgres --dbpwd HIDDEN --defaultSrsCode 2056  $file;
        done;

**Reply by ratnanil** [reply]
Word of warning however: using this approach takes **MUCH** more time than importing gdb (just a single layer TLM_BAUTEN.xtf took approx 45minutes). Maybe it is due to the fact that the all values are hard coded instead of using relational data models (i.e. storing `objektart` as integer takes much less space than storing the full name of the `objektart`).  This would definitely speak for offering the tlm dataset as a SQL Dump or a gpkg.

**Reply by ratnanil** [reply]
Just a quick benchmark. Processing the following files took 8h  😱 
- TLM_BB.xtf         
- TLM_OEV.xtf
- TLM_GEWAESSER.xtf  
- TLM_NAMEN.xtf

The amount of disk space used: 

![image](https://user-images.githubusercontent.com/12532091/143180118-fd85fceb-590b-44dc-b78a-0e87890a4881.png)

Next time around I might try gdb again (maybe by extracting the [field domains](https://gdal.org/drivers/vector/openfilegdb.html#field-domains)) and compare these values.

**Reply by edigonzales** [reply]
You probably want to disable validation with `--disableValidation`. This is something that can take much time depending on the amount of the data and complexity of the data model. But definitely something that is not done when using gdb or gpkg.

Use `--createEnumTabsWithId` when you want foreign keys instead of the enum values.

**Reply by ratnanil** [reply]
Thank you @edigonzales!

I might be sidetracking a bit, but this seems like a good place to document this workflow.. 

I'm realizing that having `objektart` *only* as text (not as integer values) is not ideal for me, since I need to `gdal_rasterize` the vector data (using `-a objektart`, which needs a numeric input). So I'm back to using the gdb datasets and stumbled on some very exiting news. As of version 3.3.0 (2021-05-03), gdal can handle field domains (see [this pr](https://github.com/OSGeo/gdal/pull/3638)). 

This means, that when using `ogr2ogr`  to export gdb to postgres, I can use the flag [`resovleDomains`](https://gdal.org/programs/ogr2ogr.html?highlight=ogr2ogr#cmdoption-ogr2ogr-resolveDomains) which creates an *additional* column `*_resolved` for every column containing a Domain. This works very nicely, HOWEVER: `objektart` is not coded via a domain, but a subtype 😱 🔪 🤦🏼 

These don't seem to be supported by gdal (yet), so I ended up extracting these with `arcpy` (see below).

And I timed importing the gdb: It took me **6 minutes to import the ENTIRE TLM Dataset** 


```
import arcpy
import os
import pandas as pd


arcpy.env.workspace = "SWISSTLM3D_2021_LV95_LN02.gdb"

datasets = arcpy.ListDatasets(feature_type='feature')
datasets = [''] + datasets if datasets is not None else []

path_list = []
for ds in datasets:
    for fc in arcpy.ListFeatureClasses(feature_dataset=ds):
        path = os.path.join(arcpy.env.workspace, ds, fc)
        path_list.append(path)


val_list = []
name_list = []
dict_list = []

for i in path_list:
    subtypes = arcpy.da.ListSubtypes(i)
    #print(subtypes)
    val_list_i = []
    name_list_i = []
    
    for key,val in subtypes.items():
        val_list_i.append(key)
        name_list_i.append(val["Name"])
    
    dict_test = {"key":val_list_i, "name": name_list_i, "path_list":i}
    dict_list.append(pd.DataFrame(dict_test))


pd.concat(dict_list).csv("subtypes.csv")
```

**Comment by KTAI-GIS** [comment]
Ich hatte das gleiche Problem. 
Mit Excel und der Power Query Funktion (nur Office 365) ist es möglich die Tabellen aus einem PDF zu extrahieren (Daten>Daten abrufen>Aus Datei>PDF). Für das SwissTLM3D habe ich Dir den Objektkatalog als Word Dokument angehängt.
[2021_04_swissTLM3D_1_9_OK_d.docx](https://github.com/GeoUnconference/discussions/files/7209357/2021_04_swissTLM3D_1_9_OK_d.docx)

**Comment by ratnanil** [comment]
vielen Dank für dein File! Das docx erlaubt auf jeden Fall noch direkteren Zugriff auf die Metadaten, aber wirklich strukturiert sind sie trotzdem noch nicht. Ich möchte festhalten, dass ich weiterhin der Meinung bin die Metadaten sollten strukturiert angeboten werden.

Oder besser noch: Dass die Daten in einem Datenformat angeboten werden wo eine solche Strukturierung _by design_ besser möglich ist (z.B. GeoPackage) wie es [der Kanton Zürich](https://opendata.swiss/de/dataset?res_format=GPKG) oder auch [Schweden](https://twitter.com/klaskarlsson/status/1439145410250031106?s=20) bereits tut.

**Reply by KTAI-GIS** [reply]
Da bin ich völlig Deiner Meinung!

**Reply by RafBov** [reply]
Hello thanks for your comment. The idea from the PDF is definitely not to be machine readeable.  In FGDB format we use subtypes and domain. In Shapefile not, everything is "flat". Geopackage could be a option for the near future but would that really solve your needs. It's just another format, which is bettser readle than fgdb in some system (and more open) . What exactly are your trying to do ? Extract data from FGDB to another format ?

**Reply by rastrau** [reply]
👍 And don't forget INTERLIS, cf. https://github.com/GeoUnconference/discussions/discussions/29#discussioncomment-1331465, @RafBov. 

But irregardless of what the format(s) in use may be able to capture or not (and various programs may be able to ingest or not), having the swissTLM3D-Objektkatalog in a machine-readable format would occasionally be attractive, in my opinion. Is it already in the backlog by any chance?

**Reply by RafBov** [reply]
it's not in the backlog, but I'll add it. As you have seen, the Interlis and the FGDB are not 100% the same. That means using Interlis to translate the FGDB won't work. Interlis is not exactly the same for different reasons. Our goal was never to get Interlis and FDBD excactly identical.  
My question is : would you better have Geopackage as new format or a machine-readeble translation of the domains/subtypes of the FGDB ?

**Reply by ratnanil** [reply]
Since nobody has answered yet, my 2 cents: I would say short term: provide subtypes and domains in a  machine readable format. Medium term: provide the data in an open format such as geopackage.

**Reply by rastrau** [reply]
Seconded.

**Reply by KTAI-GIS** [reply]
> Since nobody has answered yet, my 2 cents: I would say short term: provide subtypes and domains in a machine readable format. Medium term: provide the data in an open format such as geopackage.

I've extracted all the domains and subtypes in a XML File with an XSD but i actually stuck with FME to get something useable.
Give me some more time! Probably it's a good idea to collaborate with this to create something machine readable!

**Comment by edigonzales** [comment]
Anstelle von "Excel-Power-Query-Office-365"-Feenstaub, kann man auch einfach https://www.interlis.ch/downloads/umleditor verwenden, um aus dem INTERLIS-Modell einen HTML-Objektkatalog machen:
[fubar.html.zip](https://github.com/GeoUnconference/discussions/files/7595866/fubar.html.zip)

Falls im Modell die Attribute mit einem Kommentar dokumentiert würden, käme diese Information in der Spalte "Description". Wenn solche Modelle mit ili2db in einer DB angelegt werden, erscheinen diese Kommentare ebenfalls in der Datebank. Also alles geschenkt, sauber, transparent und automatisiert.

Unschön finde ich, dass swisstopo einen Objektkatalog nur auf Basis eines proprietären Dateiformates veröffentlicht. Also die Daten gemäss einem Datenformat beschreibt. Beispiel "TLM_Herkunft_CD": Im PDF zur GDB heisst es, dass es ein "Long Integer Coded Value" ist (Ich dachte es sei ein Aufzähltyp). Aufgelistet werden komische 100er- und 1000er- Werte. Dazu aber noch Text und eine Beschreibung des Textes. Im INTERLIS-Modell finde ich weder die Integer-Werte noch die Beschreibung. Woher kommen die?

Wahrscheinlich müsste man hier was machen, wenn man z.B. CSV-Export will:
https://github.com/claeis/umleditor/blob/master/src/ch/ehi/umleditor/objectcatalog/HtmlWriter.java
https://github.com/claeis/umleditor/blob/bf3d0a5307019f4d21984a88bf4c8bdf5eab7f4e/src/ch/ehi/umleditor/objectcatalog/ObjectCatalog.java

**Reply by rastrau** [reply]
cc @KTAI-GIS

**Reply by KTAI-GIS** [reply]
Das Interlis Modell ist suboptimal. Eine bessere Lösung sehe ich mit entsprechenden Katalogen. Da könnte man auch die domain values der gdb abbilden (9999 usw.). Ich habe auch mit FME die gdb in eine gpkg konvertiert. Der Interlisimport hat mich nicht zum Ziel geführt, hatte auch zuwenig Geduld ;-)

**Reply by rastrau** [reply]
Genau. Die numerischen Codes (bzw. die “komischen 100er- und 1000er-Werte” 😉) sind nur in der GDB (Coded Value Domain) definiert. Diese _müssten nicht_ numerisch sein (cf. https://desktop.arcgis.com/en/arcmap/latest/manage-data/geodatabases/an-overview-of-attribute-domains.htm).

Idealerweise wäre dies in Interlis gleich modelliert, mit `CatalogueObjects_V1` (IIRC). Ein Beispiel, wo das verbaut ist, ist das Modell der Sachpläne des Bundes: [`BaseModel_SectoralPlans_V1_4.ili`](http://models.geo.admin.ch/ARE/BaseModel_SectoralPlans_V1_4.ili)

**Reply by ratnanil** [reply]
> Unschön finde ich, dass swisstopo einen Objektkatalog nur auf Basis eines proprietären Dateiformates veröffentlicht.

Da stimme ich dir zu! Tatsächlich werden die Daten primär in Geodatabase produziert. Siehe dazu [Objektkatalog swissTLM3D - Ausgabe 2021](https://www.swisstopo.admin.ch/content/swisstopo-internet/de/geodata/landscape/tlm3d/_jcr_content/contentPar/tabs_copy/items/dokumente/tabPar/downloadlist/downloadItems/441_1618774195491.download/2021_04_swissTLM3D_1_9_OK_d_bf.pdf)

> ESRI Geodatabase ist das Nativformat (d.h. das Format in dem die Daten produziert werden). Der Inhalt der gelieferten Dateien ändert sich je nach Format leicht. [...] Die Beziehungen (Relationen) zwischen den Tabellen werden in der Geodatabase explizit implementiert.

**Reply by ratnanil** [reply]
> Die numerischen Codes (bzw. die “komischen 100er- und 1000er-Werte” wink) sind nur in der GDB (Coded Value Domain) definiert. 

Nicht nur in Coded Value Domain: Objektart ist als [Subtype](https://pro.arcgis.com/en/pro-app/latest/help/data/geodatabases/overview/an-overview-of-subtypes.htm) festgehalten.

**Comment by ratnanil** [comment]
> Ich habe auch mit FME die gdb in eine gpkg konvertiert

Konnte FME die Coded Value Domains auflösen? Und die Subtypes?

Wie erwähnt konnte ich die Coded Value Domains mithilfe von ogr2ogr mit dem Flag `-resolveDomains` auflösen, dabei hat ogr2ogr für jede Spalte mit einem Coded Value Domain eine zusätliche Spalte `spaltennamen_resoved` erstellt. Das ist nicht ganz optimal, aber zumindest ein Anfang. Mit Subtypes kann `ogr2ogr` meines Wissen noch gar nicht umgehen.
